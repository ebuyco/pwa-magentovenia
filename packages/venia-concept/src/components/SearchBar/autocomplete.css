.root {
    border: 1px solid rgb(var(--venia-border));
    font-size: 0.8125rem;
    box-shadow: 0 3px 10px rgba(0, 0, 0, 0.2);
    background-color: white;
    z-index: 2;
    padding: 1rem 0;
}

.statusContent {
    padding: 0.5rem 1.5rem;
}
