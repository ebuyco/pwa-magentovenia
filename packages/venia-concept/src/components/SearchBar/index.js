export { default as SearchBar } from './searchBar';
export { default } from './container';
