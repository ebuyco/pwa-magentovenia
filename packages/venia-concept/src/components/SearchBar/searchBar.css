.root {
    align-items: center;
    display: none;
    justify-content: center;
    padding: 0 1rem;
}

.root_open {
    composes: root;
    display: flex;
}

.form {
    align-items: center;
    display: grid;
    justify-items: stretch;
    margin-bottom: 0.5rem;
    width: 100%;
}

.searchInner {
       /* position: relative; */
        position: absolute;
        box-sizing: border-box;
        display: inline-flex;
        align-items: center;
        justify-content: center;
        max-width: 24rem;
        width: 100%;
        /* left: 0; */
        right: 180px;
        top: 65px;

}

.SearchAutocompleteWrapper {
    text-align: left;
    margin: -0.5rem 0 0.5rem;
    position: absolute;
    top: 100%;
    left: 0;
    right: 0;
}

@media (max-width: 400px){
    .searchInner {
        right: 70px;
        width: 26%;
        top: 56px;
    }

}
