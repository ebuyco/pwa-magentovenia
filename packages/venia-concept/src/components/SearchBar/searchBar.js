import React, { Component } from 'react';
import { Form } from 'informed';
import { bool, func, object, shape, string } from 'prop-types';
import getQueryParameterValue from '../../util/getQueryParameterValue';
import { SEARCH_QUERY_PARAMETER } from '../../RootComponents/Search/consts';
import SearchAutocomplete from './autocomplete';
import Icon from 'src/components/Icon';
import ClearIcon from 'react-feather/dist/icons/x';
import SearchIcon from 'react-feather/dist/icons/search';
import TextInput from 'src/components/TextInput';
import Trigger from 'src/components/Trigger';

import classify from 'src/classify';
import defaultClasses from './searchBar.css';

const initialValues = {
    search_query: ''
};

const clearIcon = <Icon src={ClearIcon} size={18} />;
const searchIcon = <Icon src={SearchIcon} size={18} />;

// TODO: remove export here (update story and test)
export class SearchBar extends Component {
    static propTypes = {
        classes: shape({
            clearIcon: string,
            clearIcon_off: string,
            root: string,
            searchBlock: string,
            searchBlockOpen: string,
            searchBar: string,
            searchIcon: string
        }),
        executeSearch: func.isRequired,
        history: object,
        isOpen: bool,
        location: object,
        match: object
    };

    constructor(props) {
        super(props);
        this.searchRef = React.createRef();
        this.autocompleteRef = React.createRef();
        this.state = {
            dirty: false,
            autocompleteVisible: false
        };
    }

    componentWillUnmount = () => {
        document.removeEventListener(
            'mousedown',
            this.autocompleteClick,
            false
        );
    };

    componentDidMount() {
        const searchValueFromQueryString = getQueryParameterValue({
            location: this.props.location,
            queryParameter: SEARCH_QUERY_PARAMETER
        });

        document.addEventListener('mousedown', this.autocompleteClick, false);
        this.formApi.setValue('search_query', searchValueFromQueryString);
    }

    componentDidUpdate(prevProps) {
        if (this.props.isOpen !== prevProps.isOpen) {
            this.searchRef.current.querySelector('input').focus();
        }
    }

    autocompleteClick = e => {
        if (
            this.searchRef.current.contains(e.target) ||
            this.autocompleteRef.current.contains(e.target)
        )
            return;
        this.updateAutocompleteVisible(false);
    };

    inputFocus = () => {
        this.updateAutocompleteVisible(true);
    };

    get resetButton() {
        const { resetForm, state } = this;

        return state.dirty && <Trigger action={resetForm}>{clearIcon}</Trigger>;
    }

    updateAutocompleteVisible = visible => {
        this.setState({
            autocompleteVisible: visible
        });
    };

    handleChange = ({ values }) => {
        const dirty = !!values.search_query;
        dirty !== this.state.dirty && this.setState({ dirty });
    };

    handleInputChange = () => this.updateAutocompleteVisible(true);

    handleSubmit = ({ search_query }) => {
        if (search_query) {
            this.props.executeSearch(search_query, this.props.history);
            this.updateAutocompleteVisible(false);
        }
    };

    resetForm = () => {
        this.updateAutocompleteVisible(false);
        this.formApi.reset();
    };

    setApi = formApi => {
        this.formApi = formApi;
    };

    render() {
        const { props, resetButton, formApi } = this;
        const { autocompleteVisible } = this.state;
        const { classes, isOpen } = props;
        const className = isOpen ? classes.root_open : classes.root;

        const { values } = (formApi && formApi.getState()) || {};
        const searchQuery = (values && values.search_query) || '';

        return (
            <div className={className}>
                <div className={classes.searchInner} ref={this.searchRef}>
                    <Form
                        className={classes.form}
                        getApi={this.setApi}
                        autoComplete="off"
                        initialValues={initialValues}
                        onChange={this.handleChange}
                        onSubmit={this.handleSubmit}
                    >
                        <TextInput
                            field="search_query"
                            onFocus={this.inputFocus}
                            onChange={this.handleInputChange}
                            after={resetButton}
                            before={searchIcon}
                        />

                        <div
                            className={classes.SearchAutocompleteWrapper}
                            ref={this.autocompleteRef}
                        >
                            <SearchAutocomplete
                                searchQuery={searchQuery}
                                updateAutocompleteVisible={
                                    this.updateAutocompleteVisible
                                }
                                autocompleteVisible={autocompleteVisible}
                                executeSearch={this.props.executeSearch}
                                history={this.props.history}
                            />
                        </div>
                    </Form>
                </div>
            </div>
        );
    }
}

export default classify(defaultClasses)(SearchBar);
