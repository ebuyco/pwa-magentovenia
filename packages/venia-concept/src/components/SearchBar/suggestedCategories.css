.root {
    padding: 0 1.5rem 1rem;
}

.item {
    padding-bottom: 0.72rem;
}

.item:last-child {
    padding-bottom: 0;
}
