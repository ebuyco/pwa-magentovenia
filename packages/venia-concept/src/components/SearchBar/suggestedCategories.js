import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classify from 'src/classify';
import { Link } from 'src/drivers';
import { List } from '@magento/peregrine';
import defaultClasses from './suggestedCategories.css';

class SuggestedCategories extends Component {
    static propTypes = {
        handleCategorySearch: PropTypes.func.isRequired,
        autocompleteQuery: PropTypes.string.isRequired,
        classes: PropTypes.shape({
            root: PropTypes.string,
            item: PropTypes.string
        }),
        categorySuggestions: PropTypes.arrayOf(
            PropTypes.shape({
                value_string: PropTypes.string,
                label: PropTypes.string
            })
        ).isRequired
    };

    render() {
        const {
            handleCategorySearch,
            classes,
            autocompleteQuery,
            categorySuggestions
        } = this.props;

        return (
            <List
                render="ul"
                className={classes.root}
                items={categorySuggestions}
                getItemKey={item => item.value_string}
                renderItem={({ item }) => (
                    <li className={classes.item}>
                        <Link
                            onClick={handleCategorySearch}
                            data-id={item.value_string}
                            to="/search.html"
                        >
                            <strong>{autocompleteQuery}</strong> in {item.label}
                        </Link>
                    </li>
                )}
            />
        );
    }
}

export default classify(defaultClasses)(SuggestedCategories);
