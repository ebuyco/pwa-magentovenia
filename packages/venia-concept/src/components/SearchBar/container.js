import { connect } from 'src/drivers';

import { executeSearch } from 'src/actions/app';
import SearchBar from './searchBar';

const mapStateToProps = ({ app }) => ({ app });
const mapDispatchToProps = { executeSearch };

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(SearchBar);
