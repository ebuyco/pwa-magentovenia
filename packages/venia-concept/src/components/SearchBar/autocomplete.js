import React, { Component } from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash.debounce';
import { Query } from 'src/drivers';

import classify from 'src/classify';
import { loadingIndicator } from 'src/components/LoadingIndicator';
import SuggestedCategories from './suggestedCategories';
import SuggestedProducts from './suggestedProducts';
import PRODUCT_SEARCH from '../../queries/productSearch.graphql';

import defaultClasses from './autocomplete.css';

const debounceTimeout = 200;
const suggestedCategoriesLimit = 4;
const suggestedProductsLimit = 3;

class SearchAutocomplete extends Component {
    static propTypes = {
        classes: PropTypes.shape({
            root: PropTypes.string,
            statusContent: PropTypes.string
        }),
        searchQuery: PropTypes.string.isRequired,
        autocompleteVisible: PropTypes.bool,
        executeSearch: PropTypes.func.isRequired,
        updateAutocompleteVisible: PropTypes.func.isRequired
    };

    constructor(props) {
        super(props);
        this.state = {
            isQueryUpdating: false,
            autocompleteQuery: props.searchQuery || ''
        };
    }

    componentDidUpdate = prevProps => {
        const { searchQuery } = this.props;

        if (prevProps.searchQuery !== searchQuery) {
            this.setState({ isQueryUpdating: true });
            this.updateAutocompleteQuery(searchQuery);
        }
    };

    /* Debounce this update in order to avoid multiple autocomplete query calls */
    updateAutocompleteQuery = debounce(value => {
        this.setState({
            autocompleteQuery: value,
            isQueryUpdating: false
        });
    }, debounceTimeout);

    handleCategorySearch = event => {
        event.preventDefault();
        const { id } = event.currentTarget.dataset || event.srcElement.dataset;
        this.props.updateAutocompleteVisible(false);
        this.props.executeSearch(
            this.state.autocompleteQuery,
            this.props.history,
            id
        );
    };

    handleOnProductOpen = () => this.props.updateAutocompleteVisible(false);

    // map Magento 2.3.1 schema changes to Venia 2.0.0 proptype shape to maintain backwards compatibility
    mapProducts(products) {
        return products.map(product => {
            const { small_image } = product;
            return {
                ...product,
                small_image:
                    typeof small_image === 'object'
                        ? small_image.url
                        : small_image
            };
        });
    }

    render() {
        const { classes, autocompleteVisible } = this.props;
        const { handleOnProductOpen, handleCategorySearch } = this;
        const { autocompleteQuery, isQueryUpdating } = this.state;

        if (!autocompleteVisible || autocompleteQuery.length < 3) return null;

        return (
            <Query
                query={PRODUCT_SEARCH}
                variables={{
                    inputText: autocompleteQuery
                }}
            >
                {({ loading, error, data }) => {
                    if (error)
                        return (
                            <div className={classes.root}>
                                <div className={classes.statusContent}>
                                    Data Fetch Error
                                </div>
                            </div>
                        );
                    if (loading || isQueryUpdating)
                        return (
                            <div className={classes.root}>
                                <div className={classes.statusContent}>
                                    {loadingIndicator}
                                </div>
                            </div>
                        );

                    const { filters, items } = data.products;

                    if (items.length <= 0)
                        return (
                            <div className={classes.root}>
                                <div className={classes.statusContent}>
                                    No results found, try a different search
                                </div>
                            </div>
                        );

                    const categoryFilter = filters.find(
                        filter => filter.name === 'Category'
                    );

                    const categorySuggestions = categoryFilter[
                        'filter_items'
                    ].slice(0, suggestedCategoriesLimit);

                    return (
                        <div className={classes.root}>
                            <SuggestedCategories
                                handleCategorySearch={handleCategorySearch}
                                autocompleteQuery={autocompleteQuery}
                                categorySuggestions={categorySuggestions}
                            />
                            <SuggestedProducts
                                handleOnProductOpen={handleOnProductOpen}
                                items={this.mapProducts(
                                    items.slice(0, suggestedProductsLimit)
                                )}
                            />
                        </div>
                    );
                }}
            </Query>
        );
    }
}

export default classify(defaultClasses)(SearchAutocomplete);
