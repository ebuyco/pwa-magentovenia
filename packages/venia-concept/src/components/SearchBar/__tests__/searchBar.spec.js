import React from 'react';
import TestRenderer from 'react-test-renderer';

import { Form } from 'informed';
import TextInput from 'src/components/TextInput';
import SearchBar from '../searchBar';
import SearchAutocomplete from '../autocomplete';

const removeEventListenerMock = jest.fn();
Object.defineProperty(window.document, 'removeEventListener', {
    writable: true,
    configurable: true
});
window.document.removeEventListener = removeEventListenerMock;

const buttonTypes = el => el.type === 'button';
const inputTypes = el => el.type === 'input';

const executeSearchMock = jest.fn();
const props = {
    executeSearch: executeSearchMock,
    history: {},
    location: { search: '?query=test' },
    isOpen: true
};

afterEach(() => {
    executeSearchMock.mockReset();
});

afterAll(() => {
    afterAll(() => window.document.removeEventListener.mockRestore());
});

test('renders the correct tree', () => {
    const tree = TestRenderer.create(<SearchBar {...props} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('renders the correct tree', () => {
    const newProps = {
        ...props,
        isOpen: false
    };
    const tree = TestRenderer.create(<SearchBar {...newProps} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('the input field is seeded from the location', () => {
    const expected = 'test';

    const renderer = TestRenderer.create(<SearchBar {...props} />);
    const instance = renderer.root;

    const input = instance.find(inputTypes);

    expect(input.props.value).toBe(expected);
});

test('the reset button is visible when the input field is not empty', () => {
    const renderer = TestRenderer.create(<SearchBar {...props} />);
    const instance = renderer.root;

    const buttons = instance.findAll(buttonTypes);

    expect(buttons).toHaveLength(1);
});

test('the reset button is not visible when the input field is empty', () => {
    // Force the input field to be empty.
    const testProps = {
        ...props,
        location: { search: '' }
    };

    const renderer = TestRenderer.create(<SearchBar {...testProps} />);
    const instance = renderer.root;

    const buttons = instance.findAll(buttonTypes);

    expect(buttons).toHaveLength(0);
});

test('entering text in the input causes the reset button to appear', () => {
    // Start with an empty input field.
    const testProps = {
        ...props,
        location: { search: '' }
    };

    const renderer = TestRenderer.create(<SearchBar {...testProps} />);
    const instance = renderer.root;

    // Confirm that the button is not present.
    const preChangeButtons = instance.findAll(buttonTypes);
    expect(preChangeButtons).toHaveLength(0);

    // Simulate entering text in the input field.
    const input = instance.find(inputTypes);
    input.props.onChange({
        target: { value: 'some text' }
    });

    // Test that the button appears.
    const postChangeButtons = instance.findAll(buttonTypes);
    expect(postChangeButtons).toHaveLength(1);
});

test('the reset button clears the input', () => {
    const renderer = TestRenderer.create(<SearchBar {...props} />);
    const instance = renderer.root;

    // Test that there is some text in the input to start.
    const input = instance.find(inputTypes);
    expect(input.props.value).toBe('test');

    // Simulate clicking the reset button.
    const button = instance.find(buttonTypes);
    button.props.onClick();

    // Test that the input has been cleared and autocomplete is closed.
    expect(input.props.value).toBe('');
    expect(
        instance.findByType(SearchAutocomplete).props.autocompleteVisible
    ).toBe(false);
});

test('submitting the form executes the search', () => {
    const renderer = TestRenderer.create(<SearchBar {...props} />);
    const instance = renderer.root;

    // Simulate form submit.
    const form = instance.findByType(Form);

    form.props.onSubmit({});
    expect(executeSearchMock).not.toHaveBeenCalled();

    form.props.onSubmit({
        search_query: 'test'
    });

    // Test that executeSearch was called.
    expect(executeSearchMock).toHaveBeenCalledTimes(1);
    expect(executeSearchMock).toHaveBeenNthCalledWith(1, 'test', props.history);
});

test('focusing on the text input sets state.autocompleteVisible true', () => {
    const renderer = TestRenderer.create(<SearchBar {...props} />);
    const instance = renderer.root;
    const setStateMock = jest.fn();
    instance.children[0].instance.setState = setStateMock;

    const input = instance.findByType(TextInput);
    input.props.onFocus();

    expect(setStateMock).toHaveBeenCalled();
    expect(setStateMock).toHaveBeenCalledWith({
        autocompleteVisible: true
    });
});

test('handles clicks inside and out of autocomplete input', () => {
    const renderer = TestRenderer.create(<SearchBar {...props} />);
    const instance = renderer.root;
    const setStateMock = jest.fn();

    instance.children[0].instance.searchRef.current = {};
    instance.children[0].instance.autocompleteRef.current = {};

    instance.children[0].instance.setState = setStateMock;
    instance.children[0].instance.searchRef.current.contains = jest
        .fn()
        .mockReturnValue(true);
    instance.children[0].instance.autocompleteRef.current.contains = jest
        .fn()
        .mockReturnValue(false);

    instance.children[0].instance.autocompleteClick({
        target: ''
    });

    expect(setStateMock).not.toHaveBeenCalled();

    instance.children[0].instance.searchRef.current.contains = jest
        .fn()
        .mockReturnValue(false);
    instance.children[0].instance.autocompleteRef.current.contains = jest
        .fn()
        .mockReturnValue(true);

    instance.children[0].instance.autocompleteClick({
        target: ''
    });

    expect(setStateMock).not.toHaveBeenCalled();

    instance.children[0].instance.searchRef.current.contains = jest
        .fn()
        .mockReturnValue(false);
    instance.children[0].instance.autocompleteRef.current.contains = jest
        .fn()
        .mockReturnValue(false);

    instance.children[0].instance.autocompleteClick({
        target: ''
    });

    expect(setStateMock).toHaveBeenCalledWith({
        autocompleteVisible: false
    });
});

test('removes mousedown event listener', () => {
    const renderer = TestRenderer.create(<SearchBar {...props} />);
    const instance = renderer.root;

    instance.children[0].instance.autocompleteClick = jest.fn();

    instance.children[0].instance.componentWillUnmount();

    expect(removeEventListenerMock).toHaveBeenCalledWith(
        'mousedown',
        instance.children[0].instance.autocompleteClick,
        false
    );
});
