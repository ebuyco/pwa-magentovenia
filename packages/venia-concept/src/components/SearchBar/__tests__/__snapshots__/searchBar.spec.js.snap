// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the correct tree 1`] = `
<div>
  <div>
    <form
      autoComplete="off"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <span
        style={
          Object {
            "--iconsAfter": 1,
            "--iconsBefore": 1,
          }
        }
      >
        <span>
          <input
            name="search_query"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            value="test"
          />
        </span>
        <span>
          <span>
            <svg
              fill="none"
              height={18}
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              viewBox="0 0 24 24"
              width={18}
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="11"
                cy="11"
                r="8"
              />
              <line
                x1="21"
                x2="16.65"
                y1="21"
                y2="16.65"
              />
            </svg>
          </span>
        </span>
        <span>
          <button
            onClick={[Function]}
            type="button"
          >
            <span>
              <svg
                fill="none"
                height={18}
                stroke="currentColor"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                viewBox="0 0 24 24"
                width={18}
                xmlns="http://www.w3.org/2000/svg"
              >
                <line
                  x1="18"
                  x2="6"
                  y1="6"
                  y2="18"
                />
                <line
                  x1="6"
                  x2="18"
                  y1="6"
                  y2="18"
                />
              </svg>
            </span>
          </button>
        </span>
      </span>
      <p />
      <div />
    </form>
  </div>
</div>
`;

exports[`renders the correct tree 2`] = `
<div>
  <div>
    <form
      autoComplete="off"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <span
        style={
          Object {
            "--iconsAfter": 1,
            "--iconsBefore": 1,
          }
        }
      >
        <span>
          <input
            name="search_query"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            value="test"
          />
        </span>
        <span>
          <span>
            <svg
              fill="none"
              height={18}
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              viewBox="0 0 24 24"
              width={18}
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="11"
                cy="11"
                r="8"
              />
              <line
                x1="21"
                x2="16.65"
                y1="21"
                y2="16.65"
              />
            </svg>
          </span>
        </span>
        <span>
          <button
            onClick={[Function]}
            type="button"
          >
            <span>
              <svg
                fill="none"
                height={18}
                stroke="currentColor"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                viewBox="0 0 24 24"
                width={18}
                xmlns="http://www.w3.org/2000/svg"
              >
                <line
                  x1="18"
                  x2="6"
                  y1="6"
                  y2="18"
                />
                <line
                  x1="6"
                  x2="18"
                  y1="6"
                  y2="18"
                />
              </svg>
            </span>
          </button>
        </span>
      </span>
      <p />
      <div />
    </form>
  </div>
</div>
`;
